tempesta_fw/ss_skb.c

```diff
@@ -461,136 +512,185 @@ __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
-	if (skb_shinfo(skb_dst)->nr_frags)
-		return skb_frag_address(&skb_shinfo(skb_dst)->frags[0]);
-	/* ...not found, return invalid address try next skb. */
-	return (void *)1;
+		it->ptr = skb_frag_address(&si->frags[i]);
+	return 0;
 }
 
-static void *
-__split_pgfrag(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
-	       int len)
+static int
+__split_pgfrag(struct sk_buff *skb, int i, int off, int len, TfwStr *it)
 {
 	return len > 0
-		? __split_pgfrag_add(skb, pskb, i, off, len)
-		: __split_pgfrag_del(skb, pskb, i, off, -len);
+		? __split_pgfrag_add(skb, i, off, len, it)
+		: __split_pgfrag_del(skb, i, off, -len, it);
+}
+
+static inline void *
+ss_skb_data_address(struct sk_buff *skb)
+{
+	if (skb == NULL)
+		return NULL;
+	if (skb_headlen(skb))
+		return skb->data;
+	if (skb_is_nonlinear(skb))
+		return skb_frag_address(&skb_shinfo(skb)->frags[0]);
+	BUG_ON(skb_has_frag_list(skb));
+	return NULL;
 }
 
 /**
  * Fragment @skb to add some room if @len > 0 or delete data otherwise.
  */
 static int
-__skb_fragment(struct sk_buff *skb, struct sk_buff *pskb, char *pspt,
-	       int len, TfwStr *it)
+__skb_fragment(SsSkbList *head, struct sk_buff *skb, char *pspt, int len,
+	       TfwStr *it)
 {
-	int i, dlen;
-	char *vaddr;
-	struct sk_buff *frag_i;
-
+	int ret;
+	unsigned int i, d_size, offset;
+
+	SS_DBG("[%d]: %s: in: len [%d] pspt [%p], skb [%p]: head [%p]"
+		" data [%p] tail [%p] end [%p] len [%u] data_len [%u]"
+		" truesize [%u] nr_frags [%u]\n",
+		smp_processor_id(), __func__, len, pspt, skb, skb->head,
+		skb->data, skb_tail_pointer(skb), skb_end_pointer(skb),
+		skb->len, skb->data_len, skb->truesize,
+		skb_shinfo(skb)->nr_frags);
 	BUG_ON(!len);
-	/* We can't modify data of shared or cloned skb. */
-	BUG_ON(skb_shared(skb) || skb_cloned(skb));
-
-	/* Determine where @split begins within socket buffer. */
-	dlen = skb_headlen(skb);
-	vaddr = skb->data;
-
-	SS_DBG("skb fragmentation (len=%d pspt=%p, skb: head=%p data=%p"
-	       " tail=%p end=%p len=%u data_len=%u truesize=%u"
-	       " nr_frags=%u frag_list=%p)...\n",
-	       len, pspt, skb->head, skb->data,
-	       skb_tail_pointer(skb), skb_end_pointer(skb),
-	       skb->len, skb->data_len, skb->truesize,
-	       skb_shinfo(skb)->nr_frags, skb_shinfo(skb)->frag_list);
-
-	if (pspt >= vaddr && pspt < vaddr + dlen) {
-		it->ptr = __split_linear_data(skb, pskb, pspt, len);
+
+	if ((len > (int)PAGE_SIZE) || (-len > (int)PAGE_SIZE)) {
+		SS_WARN("Attempt to add or delete too much data: %u\n", len);
+		return -EINVAL;
+	}
+
+	/*
+	 * Use @it to hold the return values from __split_pgfrag()
+	 * and __split_linear_data(). @it->ptr, @it->skb, and
+	 * @it->flags may be set to actual values. If a new SKB is
+	 * allocated, then it is stored in @it->skb. @it->ptr holds
+	 * the pointer either to data after the deleted data, or to
+	 * the area for new data. @it->flags indicates that @it->ptr
+	 * points to data in @it->skb.
+	 */
+	memset(it, 0, sizeof(TfwStr));
```

Comment:
Why do you need this? If something is wrong, then we return error code. If not, then ```it```'s members aren't used.